aiopeewee/__init__.py

```diff
@@ -30,6 +30,57 @@ def __getattr__(self, name: str) -> t.Any:
         return _ctx[name].get()
 
 
+class _transaction_async(pw._transaction):
+
+    async def __aenter__(self):
+        """Enter to async context."""
+        return super().__enter__()
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        """Exit from async context."""
+        return super().__exit__(exc_type, exc_val, exc_tb)
+
+
+class _savepoint_async(pw._savepoint):
+
+    async def __aenter__(self):
+        """Enter to async context."""
+        return super().__enter__()
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        """Exit from async context."""
+        return super().__exit__(exc_type, exc_val, exc_tb)
+
+
+class _manual_async(pw._manual):
+
+    async def __aenter__(self):
+        """Enter to async context."""
+        return super().__enter__()
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        """Exit from async context."""
+        return super().__exit__(exc_type, exc_val, exc_tb)
+
+
+class _atomic_async(pw._atomic):
+
+    async def __aenter__(self):
+        """Enter to async context."""
+        if self.db.transaction_depth() == 0:
+            args, kwargs = self._transaction_args
+            self._helper = self.db.transaction(*args, **kwargs)
+        elif isinstance(self.db.top_transaction(), pw._manual):
+            raise ValueError('Cannot enter atomic commit block while in manual commit mode.')
+        else:
+            self._helper = self.db.savepoint()
+        return await self._helper.__aenter__()
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        """Exit from async context."""
+        await self._helper.__aexit__(exc_type, exc_val, exc_tb)
```

Comment:
In peewee, nested transaction support is provided using `atomic()`, which does different things at different depth -- at 0, it creates a transaction, but at >= 1, it creates a savepoint.

```
with db.atomic():
    with db.atomic():
        thing.save() # does commit
        thing.select() # does rollback
```

Using the async-style construction with concurrent nested transactions in my earlier comment, you immediately hit `peewee.InternalError: no such savepoint`, because peewee does not guarantee the order in which savepoints are released. The release order must be in reverse, because releasing a savepoint will also release all savepoints created after it.

My naive solution was just to add async/await to `transaction()` and `savepoint()`. In my case, tasks finished in order, and savepoints were closed correctly in reverse order. A more robust solution would probably have to use `aio_event` to ensure ordering in all cases? I did have a try at that, and I can put up a PR for that if you would like, but I must confess that I am brand new to synchronization primitives.

Reply:
P.S., this definitely does not handle savepoint release ordering for arbitrary transaction depth 😅